src/transformers/models/m2m_100/convert_sonar_original_checkpoint_to_transformers.py

```diff
@@ -0,0 +1,221 @@
+    else:
+        raise ValueError("The state dict does not seem to contain SONAR encoder or decoder.")
+
+    parameter_renames = get_parameter_renames(model, state_dict, is_decoder)
+    new_state_dict = {trf_name: state_dict[fs2_name] for trf_name, fs2_name in parameter_renames.items()}
+    reorder_special_tokens(new_state_dict)
+
+    if is_decoder:
+        new_state_dict["decoder.embed_tokens.weight"] = new_state_dict["shared.weight"]
+    else:
+        new_state_dict["encoder.embed_tokens.weight"] = new_state_dict["shared.weight"]
+
+    model.load_state_dict(new_state_dict, strict=True)
+    model.tie_weights()
+
+    return model
+
+
+def test_conversion_accuracy(fairseq2_encoder_path, fairseq2_decoder_path):
+    """
+    This test is not directly invoked, because the encoder and decoder paths should be provided explicitly,
+    and these checkpoints are too heavy to download them by default, just for a test.
+    Please run the test from your code like below:
+    ```
+    from transformers.models.m2m_100.convert_sonar_original_checkpoint_to_transformers import test_conversion_accuracy
+    test_conversion_accuracy(PATH_TO_ENCODER, PATH_TO_DECODER)
+    ```
+    The fairseq2 checkpoints can be downloaded from the urls indicated in the following cards:
+        - https://github.com/facebookresearch/SONAR/blob/main/sonar/cards/text_sonar_basic_encoder.yaml
+        - https://github.com/facebookresearch/SONAR/blob/main/sonar/cards/text_sonar_basic_decoder.yaml
+
+    The reference embeddings were obtained with:
+    ```
+    from sonar.inference_pipelines.text import TextToEmbeddingModelPipeline
+    t2vec_model = TextToEmbeddingModelPipeline(encoder="text_sonar_basic_encoder", tokenizer="text_sonar_basic_encoder")
+    ref_embeddings = t2vec_model.predict(sentences, source_lang="eng_Latn")[:, :5]
+    ```
+    """
+    from transformers import NllbTokenizer
+    from transformers.modeling_outputs import BaseModelOutput
+
+    tokenizer = NllbTokenizer.from_pretrained("facebook/nllb-200-distilled-600M", clean_up_tokenization_spaces=True)
+    sentences = ["My name is SONAR.", "I can embed the sentences into vectorial space."]
+    tokenizer.src_lang = "eng_Latn"
+    batch = tokenizer(sentences, padding=True, return_tensors="pt")
+
+    print("Converting the encoder...")
+    enc = convert_sonar_checkpoint_from_disk(fairseq2_encoder_path).eval()
+    assert isinstance(enc, M2M100EncoderModel)
+
+    print("Conversion completed, testing the embedding accuracy...")
+    with torch.inference_mode():
+        enc_out = enc(**batch, pool_last_hidden_state=True)
+    assert enc_out.last_hidden_state.shape == (2, 1, 1024)
+    embeddings = enc_out.last_hidden_state.squeeze(1)
+
+    ref_embeddings = torch.tensor(
+        [[-0.005286, 0.002008, -0.000562, 0.006344, 0.006329], [-0.000330, -0.007055, 0.007644, 0.001841, 0.003727]]
+    )
+    assert torch.allclose(embeddings[:, :5], ref_embeddings, rtol=1e-3)
```

Comment:
I used the official SONAR model https://github.com/facebookresearch/SONAR/blob/main/sonar/cards/text_sonar_basic_encoder.yaml, which is the only SONAR text encoder that has ever been released so far. 

I intended this test only to reproduce how this particular model is converted (and as a template if anyone ever applies my conversion script to some models).

Reply:
I used the official SONAR model https://github.com/facebookresearch/SONAR/blob/main/sonar/cards/text_sonar_basic_encoder.yaml, which is the only SONAR text encoder that has ever been released so far.

I intended this test only to reproduce how this particular model is converted (and as a template if anyone ever applies my conversion script to some models).

Reply:
(let's remove this still, specifically because the conversion script should allow anyone that has trained a model with your framework to also convert it without hassle! We have integration tests that make sure embeddings or outputs orvalide, conversion scripts are not the place for this!)

Reply:
The function `test_conversion_accuracy` is not a part of the conversion script, it is an integration test for the conversion script. 
If you insist, though, I can remove it or move it to another file.